drivers: serial: uart_sam0: fix uart_sam0_irq_update TXC reset

pmcgaugh-daikin opened the pull request:
Replace `|=` with `=` assignment similar to the non-REV500 branch of the irq_update call.

Fixes #55386

pmcgaugh-daikin commented:
See [SAME/D5XX Datasheet](https://www.mouser.com/datasheet/2/268/60001507A-1130176.pdf) page 959 for information about the register flag that is being cleared.

nandojve commented on drivers/serial/uart_sam0.c:
```diff
@@ -905,7 +905,7 @@ static int uart_sam0_irq_update(const struct device *dev)
 	SercomUsart * const regs = config->regs;
 #if defined(SERCOM_REV500)
-	regs->INTFLAG.reg |=	SERCOM_USART_INTENCLR_ERROR
+	regs->INTFLAG.reg =	SERCOM_USART_INTENCLR_ERROR
```
Fix alignment.

```
#if defined(SERCOM_REV500)
	regs->INTFLAG.reg = SERCOM_USART_INTENCLR_ERROR
			  | SERCOM_USART_INTENCLR_RXBRK
			  | SERCOM_USART_INTENCLR_CTSIC
			  | SERCOM_USART_INTENCLR_RXS;
#else
	regs->INTFLAG.reg = SERCOM_USART_INTENCLR_RXS;
#endif
```

nandojve reviewed (changes requested):
Hi @pmcgaugh-daikin 

Thank you for the fix. However, I don't know which problem you are fixing. You need tell us following [commit-message-guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines).

If you are fixing a similar problem link this PR with the old reference to give context.This is important to us keep track about similar changes.

pmcgaugh-daikin commented on drivers/serial/uart_sam0.c:
```diff
@@ -905,10 +905,10 @@ static int uart_sam0_irq_update(const struct device *dev)
-			   |	SERCOM_USART_INTENCLR_CTSIC
-			   |	SERCOM_USART_INTENCLR_RXS;
+	regs->INTFLAG.reg =	SERCOM_USART_INTENCLR_ERROR
+			   						|	SERCOM_USART_INTENCLR_RXBRK
+			   						|	SERCOM_USART_INTENCLR_CTSIC
+			   						|	SERCOM_USART_INTENCLR_RXS;
```
I think there are issues with tab spacing, is there a standard tab space that is used and/or should zephyr drivers be using spaces for tabs instead? I was set up for 2-space tabs and autodetect tab use and it messed this spacing up. I think 4-wide tabs will fix this instance. I will update to address that.

dcpleung replied:
IIRC, tab is 8 spaces.

pmcgaugh-daikin replied:
I have completed and confirmed the fix and addressed your spacing fix request.

pmcgaugh-daikin commented:
I just discussed this update with our team again and it looks like further discussion needs to occur. Since the TXC activates the interrupt handler we need to actively clear this flag at some point. I'm not sure if this was intentional with this way the original irq_update was written, but it was keeping it working.

We were not seeing this occur because we were handling the uart_sam0_irq_tx_complete and disabling the interface when necessary and no adverse effects were occurring. Further discussion raised the possibility and we confirmed it with a test.

I think to align with the design goals of uart API for explicit acknowledged flags, we should actually be caching the TXC flag in the irq_update call, clearing it, and returning that cached value if the tx_complete is called. 

The one thing that isn't clear is the lifecycle of the cached value, is it expected that these cached values are re-cached each interrupt? (TXC, for example, would only be valid for the singular interrupt that triggered alongside the transmit complete?)

I appreciate your patience on resolving this issue.

pmcgaugh-sominos commented:
Trying to validate this in my Modbus set up, no other interfaces running but Modbus, 1 read 1 write per second (and watchdog), periodically (30 minutes) the process is locking up.
At the moment I am just caching the TXC each call of update_irq, but it seems like I am  getting stuck in a condition where the Modbus is waiting for TXC and never seeing it set (no timeout is present for modbus TX and DRE ends up tripping the watchdog, I think)
I have not tried setting DBGCTRL register, but I am seeing this without gdb attached as well. I will try this next.

pmcgaugh-daikin commented:
Trying to validate this in my Modbus set up, no other interfaces running but Modbus, 1 read 1 write per second (and watchdog), periodically (30 minutes) the process is locking up.
At the moment I am just caching the TXC each call of update_irq, but it seems like I am  getting stuck in a condition where the Modbus is waiting for TXC and never seeing it set (no timeout is present for modbus TX and DRE ends up tripping the watchdog, I think)
I have not tried setting DBGCTRL register, but I am seeing this without gdb attached as well. I will try this next.

pmcgaugh-daikin commented:
I have some additional notes: 
Getting caught after filling the fifo with the last byte and never seeing the TXC flag raise.  (cb_hander_tx in modbus_serial.c)
Disabled logging w/ no affect, turned on DBGCTRL with no affect.
Running while loop constantly (no WDT) reading reproduces the lockup very quickly
HOWEVER...adding a delay (printing to log in my case, although this could be related to interacting with UART?) just after filling the last fifo byte has the modbus running with NO issues (8 hours of no lockup at least) 

Setups attempted (not to say these are good choices, but to test some conditions):

- TXC Cached directly during, clear TXC bit in update irq (dev data always shows last cached value)
- TXC Cache latched  and only cleared on tx_complete call, clear TXC bit in update irq 
- TXC read and cleared on tx_complete call

All tests failed eventually of the same cause.

pmcgaugh-daikin commented:
Taking some measurements, on average it takes (in my particular setup...19200, 2 stop, No parity, no clock changes from default) 12 ticks before the TXC flag to set after sending the last byte.
I wrote the below algorithm and confirmed that the "delay" ticks will increment to 13, after this point the TXC seems to be reliably captured.
The only theory I HAD was that, we are receiving the TXC flag during the interrupt routine between caching and the INTFLAG clear, however, placing the TXC clear/check in the tx_complete call (I think would address this?) does not improve the issue.

``` c
static void cb_handler_tx(struct modbus_context *ctx)
{
	struct modbus_serial_config *cfg = ctx->cfg;
	int n;
	static int interupt_cnt = 99;
	static mod_enum state = MOD_IDLE;
	static int64_t tx_wait_start = 0;
	static uint8_t data_arr;
	static int delay  =0;

	if (cfg->uart_buf_ctr > 0) {
		state = MOD_STARTED;
		n = uart_fifo_fill(cfg->dev, cfg->uart_buf_ptr, cfg->uart_buf_ctr);
		data_arr=*cfg->uart_buf_ptr;
		cfg->uart_buf_ctr -= n;
		cfg->uart_buf_ptr += n;
		if ((cfg->uart_buf_ctr <= 0) && state != MOD_IDLE)
		{
			state = MOD_TXC_WAIT;
			interupt_cnt = 0;
			tx_wait_start = k_uptime_ticks();
			while (((k_uptime_ticks() - tx_wait_start) < delay)) {}
		}
		return;
	}

	/* Must wait till the transmission is complete or
	 * RS-485 transceiver could be disabled before all data has
	 * been transmitted and message will be corrupted.
	 */
	if (uart_irq_tx_complete(cfg->dev) || ((k_uptime_ticks() - tx_wait_start) > 50)) {
		if ((k_uptime_ticks() - tx_wait_start) > 50) {
			delay++;
			LOG_WRN("UP DELAY %i", delay);
		}
		/* Disable transmission */
		cfg->uart_buf_ptr = &cfg->uart_buf[0];
		modbus_serial_tx_off(ctx);
		modbus_serial_rx_on(ctx);
		//LOG_WRN("TXC! DELTA TICKS %lli",
		//  k_uptime_ticks() - tx_wait_start /*k_uptime_delta(&tx_wait_start)*/);
		state = MOD_IDLE;
	} else if (interupt_cnt == 2000 && state == MOD_TXC_WAIT) {
		LOG_WRN("LOCKUP! DELTA TICKS %lli",
			k_uptime_ticks() - tx_wait_start /*k_uptime_delta(&tx_wait_start)*/);
		while(1){}
		interupt_cnt++;
	} else if (state == MOD_TXC_WAIT) {
		interupt_cnt++;
	}
}
```

pmcgaugh-daikin commented:
I have figured out the issue and have modbus running consistently not missing any TXC flag. The issue ended  up related to caching the TXC flag first, and then importantly using that cached value to conditionally clear the TXC flag. I tried many different variations but anything that wasn't this specific combination had the risk of caching TXC=0 and then prior to execution of the INTFLAG clear line, having TXC set, and get cleared, this prevent detection of that TXC flag.

In my test cases this would lock up the modbus interface as it will not exit transmission unless tx_complete is true and we would never see the TXC for that given byte.

nandojve commented:
Hi @pmcgaugh-daikin ,

Do you believe it is possible create a specific test for this, since you invested a lot of time on it?
if yes, my suggestion is that you could add at ../tests/drivers)/uart/sam0/your_test

Make sure that you fix the compliance checks.

Thank you for the findings.

pmcgaugh-daikin commented:
Definitely @nandojve, I'll take care of all that. It might take me some time as I am trying to get this done part-time while I'm at work. I might look for some support if I get stuck but will put this in draft until I can make some progress.

nandojve commented:
Hi @pmcgaugh-daikin ,
I would like to have this fix for v3.4. Is there any news?

pmcgaugh-daikin commented:
@nandojve I have not had time to focus on this but I understand the importance. I will be working today to try and get it closed out.

pmcgaugh committed:
drivers: serial: uart_sam0: fix uart_sam0_irq_update TXC reset

drivers: serial: uart_sam0: fix uart_sam0_irq_update TXC reset bug

uart_sam0_irq_update function resets flags that will cause int. re-entry
existing implementation also clears the TXC flag if it is set
this breaks transmit complete detection

Per the SAMD5x/E5x Datasheet Sect. 34.8.6, writing '1' to the TXC will
clear the flag and disable TX complete interrupts, this should be
preserved through the irq_update for use in the tx_complete check function

The proper fix will cache the TXC value before conditionally clearing the
flag based on that cached value. If you do not condition this on the
cached value a race condition will periodically occur where
the TXC is cleared but never cached.

Fixes zephyrproject-rtos#55386

Signed-off-by: Peter McGaughey <peter.mcgaughey@daikincomfort.com>

pmcgaugh-daikin commented:
@nandojve I'm working on getting the code into good shape. I hope to have that ready today.

I spent the rest of the day trying to learn Twister and understand the testing framework and what it would entail. I am much further along than I was, but I got stuck and wanted advice for testing.

I have two boards to test with the atsame54_xpro, which uses an embedded programming chip, and a proprietary board (non-ideal?), which I use an Atmel ICE debugger to program. I see on the hardware page for Twister that `Currently only boards with support for both pyocd and nrfjprog are supported` neither of which I was able to get working...
Is there support for the Atmel ICE Debugger and OpenOCD? Is there an example? Is there a setup you can suggest for hardware testing of this code?

Any help is appreciated...I'd love to learn the testing portion for Zephyr and my team has some other drivers we are working on (TCC PWM implementation, new EEPROM Chips, and some other sensors) so we can contribute back high-quality contributions.

nandojve reviewed (dismissed):
I understood this is ready and the test can be done later, this is why I'm approving it. However, you need change status to "read to review". This is an important fix for the release and I think we should get it in.

photonthunder commented:
@pmcgaugh-daikin, I am using a SAMD51 that has an external RS485 chip with a TX/RX pin.  So in my application I enable and disable the RX and TX interrupt after every transmission.  I use uart_sam0_irq_tx_complete to make sure the last TX byte has been sent before I switch the RX/TX GPIO pin.  I need to do some more analysis, but in my case dev_data->txc_cache is always 0 so I never get a TX complete.  Looking at the code the uart_sam0_irq_update it should be holding the latest value of regs->INTFLAG.bit.TXC so I am a little confused why it isn't working.  Since you have got into the details of this I thought I would see if you had any thoughts?

One comment is that there is no #if defined(SERCOM_REV500) for txc_cache in uart_sam0_irq_tx_complete so if SERCOM_REV500 is not enabled you can never get a irq_tx_complete.
